varcode/variant_annotator.py

```diff
@@ -40,14 +40,38 @@ def describe_variant(self, variant):
 
         # dictionary from gene ID to list of transcript effects
         gene_transcript_effects_groups = {}
+
+        # mapping from Transcript objects to errors encountered
+        # while trying to annotated them
+        errors = {}
+
         for (gene_id, transcripts) in overlapping_transcript_groups.items():
             effects = []
             for transcript in transcripts:
-                effect = infer_transcript_effect(variant, transcript)
-                effects.append(effect)
+                try:
+                    effects.append(infer_transcript_effect(variant, transcript))
+                except (AssertionError, ValueError) as error:
+                    if raise_on_error:
+                        raise
+                    else:
+                        logging.warn(
+                            "Encountered error annotating %s for %s: %s",
+                            variant,
+                            transcript,
+                            error)
+                    errors[transcript] = error
```

Comment:
Yeah, it's not really possible to make a e.g. `Substitution` if you raise
an exception while trying to get the original amino acids you're
substituting, the necessary information never gets computed.

On Mon, Feb 16, 2015 at 9:35 PM, Tavi Nathanson <notifications@github.com>
wrote:

> In varcode/variant_annotator.py
> <https://github.com/hammerlab/varcode/pull/13#discussion_r24789064>:
>
> >          for (gene_id, transcripts) in overlapping_transcript_groups.items():
> >              effects = []
> >              for transcript in transcripts:
> > -                effect = infer_transcript_effect(variant, transcript)
> > -                effects.append(effect)
> > +                try:
> > +                    effects.append(infer_transcript_effect(variant, transcript))
> > +                except (AssertionError, ValueError) as error:
> > +                    if raise_on_error:
> > +                        raise
> > +                    else:
> > +                        logging.warn(
> > +                            "Encountered error annotating %s for %s: %s",
> > +                            variant,
> > +                            transcript,
> > +                            error)
> > +                    errors[transcript] = error
>
> Aha, it wasn't entirely clear to me that we wouldn't *have* a
> TranscriptEffect object because we errored in creating it.
>
> Sure, ErrorEffect could work. I don't feel strongly, though. Feel free to
> carry on :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hammerlab/varcode/pull/13/files#r24789064>.
>